### ISSUE TITLE
Centralised Rules with Annotations are Ignored

### ISSUE THREAD

**Issue by dt-atkinson** [aws/aws-xray-sdk-node#232]
Is this intended? Any rules with annotations are deliberately thrown away.

Presumably the ability to match on annotations is wanted?
https://github.com/aws/aws-xray-sdk-node/blob/8109e1e256288764571bdad45f2100c48280b8c4/packages/core/lib/middleware/sampling/service_connector.js#L115

**Comment by awssandra** [comment]
No, they are unsupported in the SDK as per documentation: https://docs.aws.amazon.com/xray/latest/devguide/xray-console-sampling.html

> 
> (optional) Attributes (key and value) – Segment attributes that are known when the sampling decision is made.
> 
> X-Ray SDK – Not supported. The SDK ignores rules that specify attributes.
> 
> Amazon API Gateway – Headers from the original HTTP request.

**Comment by dt-atkinson** [comment]
Is there a technical reason behind this, or is it just deemed not required?

**Comment by awssandra** [comment]
After re-reading the docs, I noticed we don't quite explain the intent anywhere.
So, the idea is that (currently) sampling decisions today are immutable - once something is marked as sampled it is, once something is not it is as well. With annotations, this impacts the sampling case.

The behavior of sampled vs un-sampled segments varies between the SDKs implementation as the designs diverged overtime. For Node, the segment is kept around, however for Java, a No-op segment is created and thus all the data was to be recorded is lost - I'd have to go dig into the other SDKs to see which ones do what. We have visited the idea of being able to "flip" the state of a segment in the SDKs, but there's some performance implications and use cases we're still gaining consensus on to support these scenarios and other use-cases. We do hope to have a clearer cohesive vision on the expansion of the SDK toolset in general going into 2020.

Hope that helps a bit,
Sandra
